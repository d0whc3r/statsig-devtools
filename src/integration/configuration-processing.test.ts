import { describe, expect, it } from 'vitest'

/**
 * Integration tests for configuration processing
 * These tests verify the complete flow of processing Statsig configurations
 */

describe('Configuration Processing Integration', () => {
  // Mock data that represents what we'd get from Statsig API
  const mockApiResponse = {
    feature_gates: [
      {
        id: 'gate-1',
        name: 'test_feature_gate',
        isEnabled: true,
        description: 'Test feature gate for integration testing',
        rules: [
          {
            name: 'Test Rule',
            passPercentage: 50,
            conditions: [
              {
                type: 'user_id',
                operator: 'eq',
                targetValue: 'test-user',
                field: 'userID',
              },
            ],
          },
        ],
        tags: ['test', 'integration'],
        lastModifierName: 'Test User',
        createdTime: 1640995200000, // Fixed timestamp for consistent snapshots
      },
    ],
    dynamic_configs: [
      {
        id: 'config-1',
        name: 'test_dynamic_config',
        isEnabled: true,
        description: 'Test dynamic config',
        rules: [],
        defaultValue: {
          theme: 'dark',
          maxItems: 10,
          features: ['feature1', 'feature2'],
        },
        tags: ['config'],
        lastModifierName: 'Config Manager',
        createdTime: 1640995200000,
      },
    ],
    experiments: [
      {
        id: 'exp-1',
        name: 'test_experiment',
        isEnabled: false,
        description: 'Test experiment for A/B testing',
        rules: [
          {
            name: 'Experiment Rule',
            passPercentage: 100,
            conditions: [
              {
                type: 'country',
                operator: 'in',
                targetValue: ['US', 'CA'],
                field: 'country',
              },
            ],
          },
        ],
        groups: [
          { name: 'control', size: 50, parameterValues: { variant: 'A' } },
          { name: 'treatment', size: 50, parameterValues: { variant: 'B' } },
        ],
        tags: ['experiment', 'ab-test'],
        lastModifierName: 'Experiment Manager',
        createdTime: 1640995200000,
      },
    ],
  }

  // Helper function to process configurations (simulates real processing logic)
  const processConfigurations = (apiResponse: typeof mockApiResponse) => {
    const processed = {
      totalConfigurations: 0,
      enabledConfigurations: 0,
      configurationsByType: {
        feature_gates: 0,
        dynamic_configs: 0,
        experiments: 0,
      },
      configurations: [] as any[],
    }

    // Process feature gates
    if (apiResponse.feature_gates) {
      processed.configurationsByType.feature_gates = apiResponse.feature_gates.length
      processed.totalConfigurations += apiResponse.feature_gates.length

      apiResponse.feature_gates.forEach((gate) => {
        if (gate.isEnabled) processed.enabledConfigurations++
        processed.configurations.push({
          ...gate,
          type: 'feature_gate',
          formattedName: gate.name.replace(/_/g, ' ').replace(/\b\w/g, (l) => l.toUpperCase()),
        })
      })
    }

    // Process dynamic configs
    if (apiResponse.dynamic_configs) {
      processed.configurationsByType.dynamic_configs = apiResponse.dynamic_configs.length
      processed.totalConfigurations += apiResponse.dynamic_configs.length

      apiResponse.dynamic_configs.forEach((config) => {
        if (config.isEnabled) processed.enabledConfigurations++
        processed.configurations.push({
          ...config,
          type: 'dynamic_config',
          formattedName: config.name.replace(/_/g, ' ').replace(/\b\w/g, (l) => l.toUpperCase()),
        })
      })
    }

    // Process experiments
    if (apiResponse.experiments) {
      processed.configurationsByType.experiments = apiResponse.experiments.length
      processed.totalConfigurations += apiResponse.experiments.length

      apiResponse.experiments.forEach((experiment) => {
        if (experiment.isEnabled) processed.enabledConfigurations++
        processed.configurations.push({
          ...experiment,
          type: 'experiment',
          formattedName: experiment.name.replace(/_/g, ' ').replace(/\b\w/g, (l) => l.toUpperCase()),
        })
      })
    }

    return processed
  }

  // Helper function to generate override suggestions
  const generateOverrideSuggestions = (configuration: any) => {
    const suggestions = {
      key: `statsig_override_${configuration.name}`,
      localStorage: '',
      sessionStorage: '',
      cookie: '',
    }

    switch (configuration.type) {
      case 'feature_gate':
        suggestions.localStorage = 'true'
        suggestions.sessionStorage = 'true'
        suggestions.cookie = 'enabled'
        break
      case 'dynamic_config':
        suggestions.localStorage = JSON.stringify(configuration.defaultValue || {})
        suggestions.sessionStorage = JSON.stringify({ overridden: true })
        suggestions.cookie = 'custom_config'
        break
      case 'experiment':
        suggestions.localStorage = JSON.stringify({ group: 'treatment' })
        suggestions.sessionStorage = JSON.stringify({ variant: 'B' })
        suggestions.cookie = 'treatment_group'
        break
    }

    return suggestions
  }

  describe('Configuration Processing', () => {
    it('should process API response correctly', () => {
      const processed = processConfigurations(mockApiResponse)

      expect(processed.totalConfigurations).toBe(3)
      expect(processed.enabledConfigurations).toBe(2) // gate and config are enabled
      expect(processed.configurationsByType).toEqual({
        feature_gates: 1,
        dynamic_configs: 1,
        experiments: 1,
      })
      expect(processed.configurations).toHaveLength(3)
    })

    it('should format configuration names correctly', () => {
      const processed = processConfigurations(mockApiResponse)

      const gateConfig = processed.configurations.find((c) => c.type === 'feature_gate')
      const dynamicConfig = processed.configurations.find((c) => c.type === 'dynamic_config')
      const experimentConfig = processed.configurations.find((c) => c.type === 'experiment')

      expect(gateConfig.formattedName).toBe('Test Feature Gate')
      expect(dynamicConfig.formattedName).toBe('Test Dynamic Config')
      expect(experimentConfig.formattedName).toBe('Test Experiment')
    })

    it('should handle empty API response', () => {
      const emptyResponse = {
        feature_gates: [],
        dynamic_configs: [],
        experiments: [],
      }

      const processed = processConfigurations(emptyResponse)

      expect(processed.totalConfigurations).toBe(0)
      expect(processed.enabledConfigurations).toBe(0)
      expect(processed.configurations).toHaveLength(0)
    })
  })

  describe('Override Suggestions', () => {
    it('should generate correct suggestions for feature gate', () => {
      const processed = processConfigurations(mockApiResponse)
      const gateConfig = processed.configurations.find((c) => c.type === 'feature_gate')
      const suggestions = generateOverrideSuggestions(gateConfig)

      expect(suggestions.key).toBe('statsig_override_test_feature_gate')
      expect(suggestions.localStorage).toBe('true')
      expect(suggestions.sessionStorage).toBe('true')
      expect(suggestions.cookie).toBe('enabled')
    })

    it('should generate correct suggestions for dynamic config', () => {
      const processed = processConfigurations(mockApiResponse)
      const dynamicConfig = processed.configurations.find((c) => c.type === 'dynamic_config')
      const suggestions = generateOverrideSuggestions(dynamicConfig)

      expect(suggestions.key).toBe('statsig_override_test_dynamic_config')
      expect(suggestions.localStorage).toBe(
        JSON.stringify({
          theme: 'dark',
          maxItems: 10,
          features: ['feature1', 'feature2'],
        }),
      )
      expect(suggestions.sessionStorage).toBe(JSON.stringify({ overridden: true }))
      expect(suggestions.cookie).toBe('custom_config')
    })

    it('should generate correct suggestions for experiment', () => {
      const processed = processConfigurations(mockApiResponse)
      const experimentConfig = processed.configurations.find((c) => c.type === 'experiment')
      const suggestions = generateOverrideSuggestions(experimentConfig)

      expect(suggestions.key).toBe('statsig_override_test_experiment')
      expect(suggestions.localStorage).toBe(JSON.stringify({ group: 'treatment' }))
      expect(suggestions.sessionStorage).toBe(JSON.stringify({ variant: 'B' }))
      expect(suggestions.cookie).toBe('treatment_group')
    })
  })

  describe('Integration Snapshots', () => {
    it('should match processed configuration snapshot', () => {
      const processed = processConfigurations(mockApiResponse)
      expect(processed).toMatchSnapshot('processed-configurations')
    })

    it('should match override suggestions snapshot', () => {
      const processed = processConfigurations(mockApiResponse)
      const allSuggestions = processed.configurations.map((config) => ({
        configName: config.name,
        configType: config.type,
        suggestions: generateOverrideSuggestions(config),
      }))

      expect(allSuggestions).toMatchSnapshot('override-suggestions')
    })

    it('should match complete integration flow snapshot', () => {
      const processed = processConfigurations(mockApiResponse)
      const integrationResult = {
        summary: {
          totalConfigurations: processed.totalConfigurations,
          enabledConfigurations: processed.enabledConfigurations,
          configurationsByType: processed.configurationsByType,
        },
        configurations: processed.configurations.map((config) => ({
          id: config.id,
          name: config.name,
          type: config.type,
          isEnabled: config.isEnabled,
          formattedName: config.formattedName,
          overrideSuggestions: generateOverrideSuggestions(config),
        })),
      }

      expect(integrationResult).toMatchSnapshot('complete-integration-flow')
    })
  })
})
