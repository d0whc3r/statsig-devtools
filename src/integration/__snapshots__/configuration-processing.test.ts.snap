// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Configuration Processing Integration > Integration Snapshots > should match complete integration flow snapshot > complete-integration-flow 1`] = `
{
  "configurations": [
    {
      "formattedName": "Test Feature Gate",
      "id": "gate-1",
      "isEnabled": true,
      "name": "test_feature_gate",
      "overrideSuggestions": {
        "cookie": "enabled",
        "key": "statsig_override_test_feature_gate",
        "localStorage": "true",
        "sessionStorage": "true",
      },
      "type": "feature_gate",
    },
    {
      "formattedName": "Test Dynamic Config",
      "id": "config-1",
      "isEnabled": true,
      "name": "test_dynamic_config",
      "overrideSuggestions": {
        "cookie": "custom_config",
        "key": "statsig_override_test_dynamic_config",
        "localStorage": "{"theme":"dark","maxItems":10,"features":["feature1","feature2"]}",
        "sessionStorage": "{"overridden":true}",
      },
      "type": "dynamic_config",
    },
    {
      "formattedName": "Test Experiment",
      "id": "exp-1",
      "isEnabled": false,
      "name": "test_experiment",
      "overrideSuggestions": {
        "cookie": "treatment_group",
        "key": "statsig_override_test_experiment",
        "localStorage": "{"group":"treatment"}",
        "sessionStorage": "{"variant":"B"}",
      },
      "type": "experiment",
    },
  ],
  "summary": {
    "configurationsByType": {
      "dynamic_configs": 1,
      "experiments": 1,
      "feature_gates": 1,
    },
    "enabledConfigurations": 2,
    "totalConfigurations": 3,
  },
}
`;

exports[`Configuration Processing Integration > Integration Snapshots > should match override suggestions snapshot > override-suggestions 1`] = `
[
  {
    "configName": "test_feature_gate",
    "configType": "feature_gate",
    "suggestions": {
      "cookie": "enabled",
      "key": "statsig_override_test_feature_gate",
      "localStorage": "true",
      "sessionStorage": "true",
    },
  },
  {
    "configName": "test_dynamic_config",
    "configType": "dynamic_config",
    "suggestions": {
      "cookie": "custom_config",
      "key": "statsig_override_test_dynamic_config",
      "localStorage": "{"theme":"dark","maxItems":10,"features":["feature1","feature2"]}",
      "sessionStorage": "{"overridden":true}",
    },
  },
  {
    "configName": "test_experiment",
    "configType": "experiment",
    "suggestions": {
      "cookie": "treatment_group",
      "key": "statsig_override_test_experiment",
      "localStorage": "{"group":"treatment"}",
      "sessionStorage": "{"variant":"B"}",
    },
  },
]
`;

exports[`Configuration Processing Integration > Integration Snapshots > should match processed configuration snapshot > processed-configurations 1`] = `
{
  "configurations": [
    {
      "createdTime": 1640995200000,
      "description": "Test feature gate for integration testing",
      "formattedName": "Test Feature Gate",
      "id": "gate-1",
      "isEnabled": true,
      "lastModifierName": "Test User",
      "name": "test_feature_gate",
      "rules": [
        {
          "conditions": [
            {
              "field": "userID",
              "operator": "eq",
              "targetValue": "test-user",
              "type": "user_id",
            },
          ],
          "name": "Test Rule",
          "passPercentage": 50,
        },
      ],
      "tags": [
        "test",
        "integration",
      ],
      "type": "feature_gate",
    },
    {
      "createdTime": 1640995200000,
      "defaultValue": {
        "features": [
          "feature1",
          "feature2",
        ],
        "maxItems": 10,
        "theme": "dark",
      },
      "description": "Test dynamic config",
      "formattedName": "Test Dynamic Config",
      "id": "config-1",
      "isEnabled": true,
      "lastModifierName": "Config Manager",
      "name": "test_dynamic_config",
      "rules": [],
      "tags": [
        "config",
      ],
      "type": "dynamic_config",
    },
    {
      "createdTime": 1640995200000,
      "description": "Test experiment for A/B testing",
      "formattedName": "Test Experiment",
      "groups": [
        {
          "name": "control",
          "parameterValues": {
            "variant": "A",
          },
          "size": 50,
        },
        {
          "name": "treatment",
          "parameterValues": {
            "variant": "B",
          },
          "size": 50,
        },
      ],
      "id": "exp-1",
      "isEnabled": false,
      "lastModifierName": "Experiment Manager",
      "name": "test_experiment",
      "rules": [
        {
          "conditions": [
            {
              "field": "country",
              "operator": "in",
              "targetValue": [
                "US",
                "CA",
              ],
              "type": "country",
            },
          ],
          "name": "Experiment Rule",
          "passPercentage": 100,
        },
      ],
      "tags": [
        "experiment",
        "ab-test",
      ],
      "type": "experiment",
    },
  ],
  "configurationsByType": {
    "dynamic_configs": 1,
    "experiments": 1,
    "feature_gates": 1,
  },
  "enabledConfigurations": 2,
  "totalConfigurations": 3,
}
`;
